From e8fbef49cc674ab0972f90a86455c4386b41f41f Mon Sep 17 00:00:00 2001
From: "Thulasimani,Sivakumar" <sivakumar.thulasimani@intel.com>
Date: Wed, 16 Sep 2015 10:56:09 +0530
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: Add Phy lane stagger for DP

SQUASHME! - This patch should be squashed into the following
existing patch since the original missed pulling in this part.

Author: Muley, Makarand R <makarand.r.muley@intel.com>
Date:   Tue Jun 30 12:31:44 2015 +0530
Change-Id: I697248461e03419821a18ecc40325121f37f5e25
FROM_UPSTREAM [VPG]: drm/i915: Implement chv display PHY lane stagger setup

Change-Id: I3bdfdd9e7b16fbccdb95fabf7304f268cb4c0342
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-5971
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
---
 drivers/gpu/drm/i915/intel_dp.c | 35 +++++++++++++++++++++++++++++++++--
 1 file changed, 33 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index 3877c41..7d28037 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -2949,7 +2949,7 @@ static void chv_pre_enable_dp(struct intel_encoder *encoder)
 	enum dpio_channel ch = vlv_dport_to_channel(dport);
 	int pipe = intel_crtc->pipe;
 	int data, i;
-	u32 val;
+	u32 val, stagger;
 
 	mutex_lock(&dev_priv->dpio_lock);
 
@@ -2993,7 +2993,38 @@ static void chv_pre_enable_dp(struct intel_encoder *encoder)
 	}
 
 	/* Data lane stagger programming */
-	/* FIXME: Fix up value only after power analysis */
+	if (intel_crtc->config.port_clock > 270000)
+		stagger = 0x18;
+	else if (intel_crtc->config.port_clock > 135000)
+		stagger = 0xd;
+	else if (intel_crtc->config.port_clock > 67500)
+		stagger = 0x7;
+	else if (intel_crtc->config.port_clock > 33750)
+		stagger = 0x4;
+	else
+		stagger = 0x2;
+
+	val = vlv_dpio_read(dev_priv, pipe, VLV_PCS01_DW11(ch));
+	val |= DPIO_TX2_STAGGER_MASK(0x1f);
+	vlv_dpio_write(dev_priv, pipe, VLV_PCS01_DW11(ch), val);
+
+	val = vlv_dpio_read(dev_priv, pipe, VLV_PCS23_DW11(ch));
+	val |= DPIO_TX2_STAGGER_MASK(0x1f);
+	vlv_dpio_write(dev_priv, pipe, VLV_PCS23_DW11(ch), val);
+
+	vlv_dpio_write(dev_priv, pipe, VLV_PCS01_DW12(ch),
+			DPIO_LANESTAGGER_STRAP(stagger) |
+			DPIO_LANESTAGGER_STRAP_OVRD |
+			DPIO_TX1_STAGGER_MASK(0x1f) |
+			DPIO_TX1_STAGGER_MULT(6) |
+			DPIO_TX2_STAGGER_MULT(0));
+
+	vlv_dpio_write(dev_priv, pipe, VLV_PCS23_DW12(ch),
+			DPIO_LANESTAGGER_STRAP(stagger) |
+			DPIO_LANESTAGGER_STRAP_OVRD |
+			DPIO_TX1_STAGGER_MASK(0x1f) |
+			DPIO_TX1_STAGGER_MULT(7) |
+			DPIO_TX2_STAGGER_MULT(5));
 
 	mutex_unlock(&dev_priv->dpio_lock);
 
-- 
1.9.1

